targets/simple_switch/register_access.h

```diff
@@ -0,0 +1,108 @@
+/* Copyright 2019-present Cisco Systems, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Andy Fingerhut (jafinger@cisco.com)
+ *
+ */
+
+#ifndef SIMPLE_SWITCH_REGISTER_ACCESS_H_
+#define SIMPLE_SWITCH_REGISTER_ACCESS_H_
+
+#define PACKET_LENGTH_REG_IDX             0
+
+#define CLONE_MIRROR_SESSION_ID_REG_IDX   1
+#define CLONE_MIRROR_SESSION_ID_MASK      0x000000000000ffff
+#define CLONE_MIRROR_SESSION_ID_SHIFT     0
+#define CLONE_FIELD_LIST_REG_IDX          1
+#define CLONE_FIELD_LIST_MASK             0x00000000ffff0000
+#define CLONE_FIELD_LIST_SHIFT            16
+#define LF_FIELD_LIST_REG_IDX             1
+#define LF_FIELD_LIST_MASK                0x0000ffff00000000
+#define LF_FIELD_LIST_SHIFT               32
+#define RESUBMIT_FLAG_REG_IDX             1
+#define RESUBMIT_FLAG_MASK                0xffff000000000000
+#define RESUBMIT_FLAG_SHIFT               48
+
+#define RECIRCULATE_FLAG_REG_IDX          2
+#define RECIRCULATE_FLAG_MASK             0x000000000000ffff
+#define RECIRCULATE_FLAG_SHIFT            0
+
+
+class RegisterAccess {
+ public:
+    static void clear_all(Packet &pkt) {
```

Comment:
let's use a `Packet` pointer for all the functions which modify the packet.

Reply:
Here is where my C++ ignorance shows again.  When I use packet pointers, there are places where I attempt to call one of these functions that give an error message I'm not sure of the best way to address. Example below.  What would you recommend for this?

I change the definition of  `get_clone_mirror_session_id` to this in register_access.h, to use a pointer to Packet instead of a reference:
```
    static uint16_t get_clone_mirror_session_id(Packet *pkt) {
        uint64_t rv = pkt->get_register(CLONE_MIRROR_SESSION_ID_REG_IDX);
        return (uint16_t) ((rv & CLONE_MIRROR_SESSION_ID_MASK) >>
                           CLONE_MIRROR_SESSION_ID_SHIFT);
    }
```

In simple_switch.cpp:
```
void
SimpleSwitch::ingress_thread() {

// ... later ...

    unsigned int clone_mirror_session_id =
        RegisterAccess::get_clone_mirror_session_id(packet);
```
when compiling I get this error message that I would love recomendations for fixing:
```
simple_switch.cpp: In member function ‘void SimpleSwitch::ingress_thread()’:
simple_switch.cpp:505:59: error: no matching function for call to ‘RegisterAccess::get_clone_mirror_session_id(std::unique_ptr<bm::Packet>&)’
         RegisterAccess::get_clone_mirror_session_id(packet);
                                                           ^
In file included from simple_switch.cpp:36:0:
register_access.h:51:21: note: candidate: static uint16_t RegisterAccess::get_clone_mirror_session_id(bm::Packet*)
     static uint16_t get_clone_mirror_session_id(Packet *pkt) {
                     ^
```

Reply:
With latest commit, this comment is the only one that I have not addressed yet.  Open to suggestions on this one on best way to address it.

Reply:
When packet is a `std::unique_ptr`, and you call a function that doesn't take ownership of the pointer, but mutates the pointee, just pass a raw pointer to the function. So in your case, the simple_switch.cpp code should be:
```
auto clone_mirror_session_id =
        RegisterAccess::get_clone_mirror_session_id(packet.get());
```
https://en.cppreference.com/w/cpp/memory/unique_ptr/get

Reply:
Done.  Thanks for the coding advice!